File: src/worker-parser-rust/src/azure/event_grid/consumer.rs
//! Event Grid Namespace pull-delivery consumer.
//!
//! Mirrors the Python `shared/event_consumer.py` — receives, acknowledges, and
//! releases events from an Event Grid Namespace subscription using the REST API.

use std::time::Duration;

use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::azure::credential::ManagedIdentityCredential;

/// Event Grid resource scope for token acquisition.
const EVENTGRID_RESOURCE: &str = "https://eventgrid.azure.net/";

/// Event Grid Namespace API version.
const API_VERSION: &str = "2024-06-01";

/// HTTP request timeout for Event Grid operations.
const REQUEST_TIMEOUT: Duration = Duration::from_secs(60);

/// A single received event with its broker metadata.
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReceiveDetails {
    pub broker_properties: BrokerProperties,
    pub event: CloudEvent,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BrokerProperties {
    pub lock_token: String,
}

/// CloudEvents v1.0 envelope — just the fields we need.
#[derive(Debug, Clone, Deserialize)]
pub struct CloudEvent {
    pub id: String,
    #[serde(rename = "type")]
    pub event_type: String,
    pub source: String,
    #[serde(default)]
    pub subject: Option<String>,
    #[serde(default)]
    pub data: Option<Value>,
}

#[derive(Deserialize)]
struct ReceiveResponse {
    details: Vec<ReceiveDetails>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct LockTokensBody {
    lock_tokens: Vec<String>,
}

/// Async Event Grid Namespace pull-delivery consumer.
#[derive(Clone)]
pub struct EventGridConsumer {
    client: Client,
    credential: ManagedIdentityCredential,
    endpoint: String,
    topic: String,
    subscription: String,
}

impl EventGridConsumer {
    pub fn new(
        endpoint: String,
        topic: String,
        subscription: String,
        credential: ManagedIdentityCredential,
    ) -> Self {
        Self {
            client: Client::builder()
                .timeout(REQUEST_TIMEOUT)
                .build()
                .expect("failed to build HTTP client"),
            credential,
            endpoint: endpoint.trim_end_matches('/').to_owned(),
            topic,
            subscription,
        }
    }

    /// Pull a batch of events from the subscription.
    pub async fn receive_events(
        &self,
        max_events: u32,
        max_wait_time_secs: u32,
    ) -> Result<Vec<ReceiveDetails>, EventGridConsumerError> {
        let token = self.get_token().await?;
        let url = format!(
            "{}/topics/{}/eventsubscriptions/{}:receive?api-version={API_VERSION}&maxEvents={max_events}&maxWaitTime={max_wait_time_secs}",
            self.endpoint, self.topic, self.subscription
        );

        let resp = self
            .client
            .post(&url)
            .header("Authorization", format!("Bearer {token}"))
            .header("Content-Type", "application/json")
            .body("{}")
            .send()
            .await
            .map_err(|e| EventGridConsumerError::Request(e.to_string()))?;

        if !resp.status().is_success() {
            let status = resp.status();
            let body = resp.text().await.unwrap_or_default();
            return Err(EventGridConsumerError::Request(format!(
                "receive HTTP {status}: {body}"
            )));
        }

        let body: ReceiveResponse = resp
            .json()
            .await
            .map_err(|e| EventGridConsumerError::Request(e.to_string()))?;

        Ok(body.details)
    }

    /// Acknowledge events so they are removed from the subscription.
    pub async fn acknowledge(
        &self,
        lock_tokens: Vec<String>,
    ) -> Result<(), EventGridConsumerError> {
        if lock_tokens.is_empty() {
            return Ok(());
        }
        self.lock_token_action("acknowledge", lock_tokens).await
    }

    /// Release events back to the subscription for redelivery.
    pub async fn release(
        &self,
        lock_tokens: Vec<String>,
    ) -> Result<(), EventGridConsumerError> {
        if lock_tokens.is_empty() {
            return Ok(());
        }
        self.lock_token_action("release", lock_tokens).await
    }

    async fn lock_token_action(
        &self,
        action: &str,
        lock_tokens: Vec<String>,
    ) -> Result<(), EventGridConsumerError> {
        let token = self.get_token().await?;
        let url = format!(
            "{}/topics/{}/eventsubscriptions/{}:{action}?api-version={API_VERSION}",
            self.endpoint, self.topic, self.subscription
        );

        let body = LockTokensBody { lock_tokens };
        let resp = self
            .client
            .post(&url)
            .header("Authorization", format!("Bearer {token}"))
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await
            .map_err(|e| EventGridConsumerError::Request(e.to_string()))?;

        if !resp.status().is_success() {
            let status = resp.status();
            let body_text = resp.text().await.unwrap_or_default();
            tracing::warn!(
                action,
                status = %status,
                "{action} partial failure: {body_text}"
            );
        }

        Ok(())
    }

    async fn get_token(&self) -> Result<String, EventGridConsumerError> {
        self.credential
            .get_token(EVENTGRID_RESOURCE)
            .await
            .map_err(|e| EventGridConsumerError::Auth(e.to_string()))
    }
}

#[derive(Debug, thiserror::Error)]
pub enum EventGridConsumerError {
    #[error("authentication error: {0}")]
    Auth(String),
    #[error("event grid request failed: {0}")]
    Request(String),
}

File: src/worker-graph-builder-go/internal/azure/eventgrid/consumer.go
// Package eventgrid provides Event Grid Namespace pull-delivery consumer and publisher.
package eventgrid

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"time"

	"github.com/christopherhouse/integrisight/worker-graph-builder-go/internal/azure/credential"
)

const (
	eventGridResource   = "https://eventgrid.azure.net/"
	apiVersion          = "2024-06-01"
	consumerTimeout     = 60 * time.Second
)

// CloudEvent is a received CloudEvents v1.0 envelope.
type CloudEvent struct {
	ID         string            `json:"id"`
	Type       string            `json:"type"`
	Source     string            `json:"source"`
	Subject    string            `json:"subject,omitempty"`
	Data       json.RawMessage   `json:"data,omitempty"`
	Extensions map[string]any    `json:"extensions,omitempty"`
}

// BrokerProperties contains Event Grid broker metadata for a received event.
type BrokerProperties struct {
	LockToken string `json:"lockToken"`
}

// ReceiveDetails bundles a CloudEvent with its broker metadata.
type ReceiveDetails struct {
	BrokerProperties BrokerProperties `json:"brokerProperties"`
	Event            CloudEvent       `json:"event"`
}

type receiveResponse struct {
	Details []ReceiveDetails `json:"details"`
}

type lockTokensBody struct {
	LockTokens []string `json:"lockTokens"`
}

// EventGridConsumer pulls events from an Event Grid Namespace subscription.
type EventGridConsumer struct {
	client       *http.Client
	credential   *credential.ManagedIdentityCredential
	endpoint     string
	topic        string
	subscription string
}

// NewConsumer creates an EventGridConsumer.
func NewConsumer(endpoint, topic, subscription string, cred *credential.ManagedIdentityCredential) *EventGridConsumer {
	return &EventGridConsumer{
		client:       &http.Client{Timeout: consumerTimeout},
		credential:   cred,
		endpoint:     trimSlash(endpoint),
		topic:        topic,
		subscription: subscription,
	}
}

// ReceiveEvents pulls up to maxEvents events, waiting up to maxWaitSecs seconds.
func (c *EventGridConsumer) ReceiveEvents(ctx context.Context, maxEvents, maxWaitSecs int) ([]ReceiveDetails, error) {
	token, err := c.getToken()
	if err != nil {
		return nil, err
	}

	url := fmt.Sprintf(
		"%s/topics/%s/eventsubscriptions/%s:receive?api-version=%s&maxEvents=%d&maxWaitTime=%d",
		c.endpoint, c.topic, c.subscription, apiVersion, maxEvents, maxWaitSecs,
	)

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewBufferString("{}"))
	if err != nil {
		return nil, fmt.Errorf("build receive request: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("receive request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("receive HTTP %d: %s", resp.StatusCode, body)
	}

	var rr receiveResponse
	if err := json.NewDecoder(resp.Body).Decode(&rr); err != nil {
		return nil, fmt.Errorf("decode receive response: %w", err)
	}
	return rr.Details, nil
}

// Acknowledge removes the given events from the subscription.
func (c *EventGridConsumer) Acknowledge(ctx context.Context, lockTokens []string) error {
	if len(lockTokens) == 0 {
		return nil
	}
	return c.lockTokenAction(ctx, "acknowledge", lockTokens)
}

// Release returns the events to the subscription for redelivery.
func (c *EventGridConsumer) Release(ctx context.Context, lockTokens []string) error {
	if len(lockTokens) == 0 {
		return nil
	}
	return c.lockTokenAction(ctx, "release", lockTokens)
}

func (c *EventGridConsumer) lockTokenAction(ctx context.Context, action string, lockTokens []string) error {
	token, err := c.getToken()
	if err != nil {
		return err
	}

	url := fmt.Sprintf(
		"%s/topics/%s/eventsubscriptions/%s:%s?api-version=%s",
		c.endpoint, c.topic, c.subscription, action, apiVersion,
	)

	body := lockTokensBody{LockTokens: lockTokens}
	bodyBytes, err := json.Marshal(body)
	if err != nil {
		return fmt.Errorf("marshal lock tokens: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(bodyBytes))
	if err != nil {
		return fmt.Errorf("build %s request: %w", action, err)
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.client.Do(req)
	if err != nil {
		return fmt.Errorf("%s request: %w", action, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		respBody, _ := io.ReadAll(resp.Body)
		// Non-fatal: partial failures are best-effort; log and continue.
		slog.Warn("eventgrid_lock_token_action_partial_failure",
			"action", action,
			"status", resp.StatusCode,
			"body", string(respBody),
		)
	}
	return nil
}

func (c *EventGridConsumer) getToken() (string, error) {
	tok, err := c.credential.GetToken(eventGridResource)
	if err != nil {
		return "", fmt.Errorf("eventgrid consumer auth: %w", err)
	}
	return tok, nil
}

func trimSlash(s string) string {
	for len(s) > 0 && s[len(s)-1] == '/' {
		s = s[:len(s)-1]
	}
	return s
}
